File: wifi/CommClient-Server/app/schemas/channel.py
"""
Channel / room schemas.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class ChannelCreate(BaseModel):
    type: str = Field(..., pattern=r"^(dm|group)$")
    name: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = Field(None, max_length=1000)
    member_ids: list[str] = Field(default_factory=list)


class ChannelUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=128)
    description: str | None = Field(None, max_length=1000)
    avatar_url: str | None = None


class ChannelMemberInfo(BaseModel):
    user_id: str
    username: str
    display_name: str
    avatar_url: str | None
    status: str
    role: str
    joined_at: datetime


class ChannelResponse(BaseModel):
    id: str
    type: str
    name: str | None
    description: str | None
    avatar_url: str | None
    created_by: str | None
    is_active: bool
    members: list[ChannelMemberInfo]
    member_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ChannelListResponse(BaseModel):
    channels: list[ChannelResponse]
    total: int


class AddMemberRequest(BaseModel):
    user_id: str
    role: str = Field("member", pattern=r"^(admin|moderator|member)$")


class ChannelBrief(BaseModel):
    id: str
    type: str
    name: str | None
    member_count: int
    last_message_at: datetime | None = None

    class Config:
        from_attributes = True


# ── Per-user channel preferences ─────────────────────────

class ChannelArchiveRequest(BaseModel):
    archived: bool


class ChannelPinRequest(BaseModel):
    pinned: bool


class ChannelMuteRequest(BaseModel):
    muted: bool
    mute_until: datetime | None = None


class ChannelReadRequest(BaseModel):
    message_id: str | None = None


class ChannelMemberPrefsResponse(BaseModel):
    channel_id: str
    user_id: str
    is_archived: bool
    is_pinned: bool
    is_muted: bool
    mute_until: datetime | None
    last_read_at: datetime | None
    last_read_message_id: str | None

    class Config:
        from_attributes = True

File: wifi/Helen-Mobile-Flutter/lib/data/models/channel.dart
/// Channel / conversation model.
library;

import 'package:freezed_annotation/freezed_annotation.dart';

part 'channel.freezed.dart';
part 'channel.g.dart';

@freezed
class Channel with _$Channel {
  const factory Channel({
    required String id,
    required String type, // direct | group | public | system
    String? name,
    String? description,
    @JsonKey(name: 'owner_id') String? ownerId,
    @JsonKey(name: 'last_message_id') String? lastMessageId,
    @JsonKey(name: 'last_message_at') DateTime? lastMessageAt,
    @JsonKey(name: 'unread_count') @Default(0) int unreadCount,
    @JsonKey(name: 'member_count') @Default(0) int memberCount,
    @Default(<ChannelMember>[]) List<ChannelMember> members,
    @JsonKey(name: 'is_pinned') @Default(false) bool isPinned,
    @JsonKey(name: 'is_archived') @Default(false) bool isArchived,
    @JsonKey(name: 'slow_mode_seconds') @Default(0) int slowModeSeconds,
    @JsonKey(name: 'ttl_seconds') @Default(0) int ttlSeconds,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Channel;

  factory Channel.fromJson(Map<String, dynamic> json) => _$ChannelFromJson(json);
}

@freezed
class ChannelMember with _$ChannelMember {
  const factory ChannelMember({
    @JsonKey(name: 'user_id') required String userId,
    @Default('member') String role,
    @JsonKey(name: 'joined_at') DateTime? joinedAt,
    @JsonKey(name: 'last_read_at') DateTime? lastReadAt,
  }) = _ChannelMember;

  factory ChannelMember.fromJson(Map<String, dynamic> json) =>
      _$ChannelMemberFromJson(json);
}
